lua/weapons/gmod_tool/stools/selfchanger.lua

```diff
@@ -1,93 +1,120 @@
+end
 
 function TOOL:sendSetVal(ent, var, val)
-	--Msg(string.format("value at ssi = %f\n", value))
-	ent[var] = val
-	local allPlayers = RecipientFilter()
-	allPlayers:AddAllPlayers()
-	umsg.Start("umsgSelfNewSet", allPlayers)	--do we need to send entity with all user messages (so we know which pannel we are talking about?)
-		umsg.Entity(ent)
-		umsg.String(var)
-		umsg.Float(val)
-	umsg.End() 
-end
+	local bnv = (self:GetClientNumber("widvaluenum") ~= 0)
 
-TOOL.ClientConVar["widvariable"] = ""
-TOOL.ClientConVar["widvalue"] = ""
+	if(bnv) then
+		ent[var] = (tonumber(val) or 0)
+	else
+		ent[var] = (tostring(val) or "")
+	end
 
+	playerNotify(var, val, (ent and ent[var] or "N/A"), self:GetOwner())
+
+	net.Start("selfchangerSetEntitySetting")
+		net.WriteEntity(ent)
+		net.WriteString(var)
+		net.WriteString(val)
+		net.WriteBool(bnv)
+	net.Broadcast()
+end
 
 function TOOL:LeftClick( trace )
 	if trace.Entity && trace.Entity:IsPlayer() then return false end
 	if (CLIENT) then return true end
-	--if (SERVER) then return true end
-	
+
+	local widValue    = self:GetClientInfo("widvalue")
 	local widVariable = self:GetClientInfo("widvariable")
-	local widValue = self:GetClientInfo("widvalue")
-		
+
 	--if (trace.Entity:IsValid() && trace.Entity.guiPanelVersion && trace.Entity.pl == ply) then
 	--trace.Entity[widVariable] = widValue
 	self:sendSetVal(trace.Entity, widVariable, widValue)
-	Msg("done\n")
-		--return true
-	--end
-	
+
 	return true
 end
 
 function TOOL:RightClick( trace )
 	if trace.Entity && trace.Entity:IsPlayer() then return false end
 	--if (CLIENT) then return true end
 	if (SERVER) then return true end
-	
+
 	--local themeName = self:GetClientInfo("themename")
-		
+
 	--if (trace.Entity:IsValid() && trace.Entity.guiPanelVersion && trace.Entity.pl == ply) then
 		--trace.Entity:SetPanelScheme(themeName)
 		--return true
 	--end
 	gpSetupVals(trace.Entity)
-	
+
 	return true
 end
 
 
 function TOOL:Think()
 end
 
-function TOOL.BuildCPanel(panel)
-
-	panel:AddControl("Header", { Text = "#Tool_wire_modular_panel_name", Description = "#Tool_wire_modular_panel_desc" })
-	
-	panel:AddControl("TextBox", {Label = "#Tool_selfchanger_widvariable", MaxLength = tostring(50), Command = "selfchanger_widvariable"})
-	panel:AddControl("TextBox", {Label = "#Tool_selfchanger_widvalue", MaxLength = tostring(50), Command = "selfchanger_widvalue"})
-	--panel:AddControl("TextBox", {Label = "#Tool_selfchanger_widproperty", MaxLength = tostring(50), Command = "selfchanger_widproperty"})
-	--panel:AddControl("TextBox", {Label = "#Tool_selfchanger_themename", MaxLength = tostring(50), Command = "selfchanger_themename"})
-end
\ No newline at end of file
+function TOOL.BuildCPanel(cPanel)
+	local pItem -- pItem is the current panel created
+	pItem = cPanel:SetName("#tool.selfchanger.name")
+	pItem = cPanel:Help   ("#tool.selfchanger.desc")
+	pItem = cPanel:TextEntry("#tool.selfchanger.widvariable", "selfchanger_widvariable")
```

Comment:
Correct me if I'm wrong, but surely you don't need to store the returned panel object for each of these just to set the tooltip, right? I mean can't you do, for example: `cPanel:TextEntry("Foo"):SetTooltip("Bar")`?

So it would become something like:

```lua
cPanel:SetName("#tool.selfchanger.name")
cPanel:Help("#tool.selfchanger.desc")
cPanel:TextEntry("#tool.selfchanger.widvariable", "selfchanger_widvariable"):SetTooltip("#tool.selfchanger.widvariable")
cPanel:TextEntry("#tool.selfchanger.widvalue", "selfchanger_widvalue"):SetTooltip("#tool.selfchanger.widvalue")
cPanel:CheckBox("#tool.selfchanger.widvaluenum", "selfchanger_widvaluenum"):SetTooltip("#tool.selfchanger.widvaluenum")
```

Reply:
True, but it is longer. I will change it if you insist.